geo/src/algorithm/relate/mod.rs

```diff
@@ -0,0 +1,111 @@
+pub(crate) use edge_end_builder::EdgeEndBuilder;
+pub use geomgraph::intersection_matrix::IntersectionMatrix;
+
+use crate::{
+    GeoFloat, Geometry, GeometryCow, Line, LineString, MultiLineString, MultiPoint, MultiPolygon,
+    Point, Polygon, Rect, Triangle,
+};
+
+mod edge_end_builder;
+mod geomgraph;
+mod relate_operation;
+
+/// Topologically relate two geometries based on [DE-9IM](https://en.wikipedia.org/wiki/DE-9IM) semantics.
+///
+/// See [`IntersectionMatrix`] for details.
+/// ```
+/// use geo::{Coordinate, Line, Rect, line_string};
+/// use crate::geo::relate::Relate;
+///
+/// let line = Line::new(Coordinate { x: 2.0, y: 2.0}, Coordinate { x: 4.0, y: 4.0 });
+/// let rect = Rect::new(Coordinate { x: 2.0, y: 2.0}, Coordinate { x: 4.0, y: 4.0 });
+/// let intersection_matrix = rect.relate(&line);
+///
+/// assert!(intersection_matrix.is_intersects());
+/// assert!(!intersection_matrix.is_disjoint());
+/// assert!(intersection_matrix.is_contains());
+/// assert!(!intersection_matrix.is_within());
+///
+/// let line = Line::new(Coordinate { x: 1.0, y: 1.0}, Coordinate { x: 5.0, y: 5.0 });
+/// let rect = Rect::new(Coordinate { x: 2.0, y: 2.0}, Coordinate { x: 4.0, y: 4.0 });
+/// let intersection_matrix = rect.relate(&line);
+/// assert!(intersection_matrix.is_intersects());
+/// assert!(!intersection_matrix.is_disjoint());
+/// assert!(!intersection_matrix.is_contains());
+/// assert!(!intersection_matrix.is_within());
+///
+/// let rect_boundary = line_string![
+///     (x: 2.0, y: 2.0),
+///     (x: 4.0, y: 2.0),
+///     (x: 4.0, y: 4.0),
+///     (x: 2.0, y: 4.0),
+///     (x: 2.0, y: 2.0)
+/// ];
+/// let intersection_matrix = rect.relate(&rect_boundary);
+/// assert!(intersection_matrix.is_intersects());
+/// assert!(!intersection_matrix.is_disjoint());
+/// // According to DE-9IM, polygons don't contain their own boundary
+/// assert!(!intersection_matrix.is_contains());
+/// assert!(!intersection_matrix.is_within());
+/// ```
+///
+/// Note: `Relate` must not be called on geometries containing `NaN` coordinates.
+pub trait Relate<F, T> {
+    fn relate(&self, other: &T) -> IntersectionMatrix;
+}
+
+impl<F: GeoFloat> Relate<F, GeometryCow<'_, F>> for GeometryCow<'_, F> {
+    fn relate(&self, other: &GeometryCow<F>) -> IntersectionMatrix {
+        let mut relate_computer = relate_operation::RelateOperation::new(self, other);
+        relate_computer.compute_intersection_matrix()
+    }
+}
+
+macro_rules! relate_impl {
+    ($(($k:ty, $t:ty),)*) => {
+        $(
+            impl<F: GeoFloat> Relate<F, $t> for $k {
+                fn relate(&self, other: &$t) -> IntersectionMatrix {
+                    GeometryCow::from(self).relate(&GeometryCow::from(other))
```

Comment:
FYI: this line is why I introduced `GeometryCow`.

Unlike much of our other code, the GeometryGraph was implemented to deal only with `Geometry`, rather than have a bunch of different implementations for each Geometry variant.

However, if we want to be able to call something like: 

```
multi_line_string.relate(polygon)
```

We'd actually have to do something like:
```
Geometry::from(multi_line_string.clone()).relate(Geometry::from(polygon.clone()))
```

Those clones are really unfortunate - hence `GeometryCow`, which is an enum like `Geometry` but can be made from a variant without cloning.